### PR TITLE
Fix substitution filtering

### PR DESCRIPTION
**Pull request by dwightguth**

### CODE REVIEW

**Comment by ttuegel** [comment]
This fixed my problem with the Haskell backend tests.

File: lib/ast/AST.cpp [hunk 1]
```diff
@@ -945,8 +945,11 @@ sptr<KOREPattern> KORECompositePattern::filterSubstitution(PrettyPrintData const
           return result->getArguments()[0];
         }
       }
+      return result;
+    } else if (constructor->getName() == "\\or") {
+      return result;
     }
-    return result;
+    return shared_from_this();
```
Comment (theo25) [inline lib/ast/AST.cpp:956]: So, if I am reading this correctly, there is a case when you do recurse but then ignore the result returned from the recursion (when after the for loop at line 934, you don't enter the `if` or the `else if` blocks). This seems off but if it is intended, let me know and I will approve.

**Review by theo25 — CHANGES_REQUESTED** [review]
I left a comment for you.

File: lib/ast/AST.cpp (hunk 1 above)
Comment (dwightguth) [reply lib/ast/AST.cpp:956]: I will rewrite this so that we skip the recursion in this case.

**Review by theo25 — APPROVED** [review]
Looks good to me.

**Pull request merged**